gensim/models/keyedvectors.py

```diff
@@ -158,264 +163,425 @@
-        return len(self.closer_than(entity1, entity2)) + 1
+    # backward compatibility
+    @property
+    def vectors_norm(self):
+        self.fill_norms()
+        return self.vectors / self.norms[..., np.newaxis]
 
+    def fill_norms(self, force=False):
+        """
+        Ensure per-vector norms are available.
 
-class WordEmbeddingsKeyedVectors(BaseKeyedVectors):
-    """Class containing common methods for operations over word vectors."""
-    def __init__(self, vector_size):
-        super(WordEmbeddingsKeyedVectors, self).__init__(vector_size=vector_size)
-        self.vectors_norm = None
-        self.index2word = []
+        (Any code which modifies vectors should ensure the
+        accompanying norms are recalculated, or 'None'-out
+        'norms' to trigger full recalc later.)
+        """
+        if self.norms is None or force:
+            self.norms = np.linalg.norm(self.vectors, axis=1)
 
-    @property
-    @deprecated("Attribute will be removed in 4.0.0, use self instead")
-    def wv(self):
-        return self
+    @vectors_norm.setter
+    def vectors_norm(self, _):
+        pass  # no-op; shouldn't be set
 
     @property
     def index2entity(self):
-        return self.index2word
+        return self.index_to_key
 
     @index2entity.setter
     def index2entity(self, value):
-        self.index2word = value
+        self.index_to_key = value
 
     @property
-    @deprecated("Attribute will be removed in 4.0.0, use self.vectors instead")
-    def syn0(self):
-        return self.vectors
+    def index2word(self):
+        return self.index_to_key
 
-    @syn0.setter
-    @deprecated("Attribute will be removed in 4.0.0, use self.vectors instead")
-    def syn0(self, value):
-        self.vectors = value
+    @index2word.setter
+    def index2word(self, value):
+        self.index_to_key = value
 
     @property
-    @deprecated("Attribute will be removed in 4.0.0, use self.vectors_norm instead")
-    def syn0norm(self):
-        return self.vectors_norm
+    def vocab(self):
+        raise NotImplementedError(
+            "The .vocab dict of 'Vocab' propery objects, one per key, has been removed.\n"
+            "See the KeyedVectors .key_to_index dict, .index_to_key list, and methods\n"
+            ".get_vecattr(key, attr)/.set_vecattr(key, attr, new_val) for replacement\n"
+            "functionality."
+        )
+
+    @vocab.setter
+    def vocab(self, value):
+        self.vocab()  # trigger above NotImplementedError
 
-    @syn0norm.setter
-    @deprecated("Attribute will be removed in 4.0.0, use self.vectors_norm instead")
-    def syn0norm(self, value):
-        self.vectors_norm = value
+    @property
+    def pseudovocab(self):
+        """ pseudodict providing pseudovocab objects
 
-    def __contains__(self, word):
-        return word in self.vocab
+        not efficient, temp backcompat workaround 'just in case' a .vocab use can't adapt
```

Comment:
What?

Reply:
@gojomo is this class still necessary? I have no idea what I just read in this docstring.

Reply:
This was from a workaround/testing-kludge that could make the new 'vecattr' mechanism simulate the old `Vocab` objects, in case any older usages couldn't adapt. It's unlikely worth the complexity, so will be removed. (Further writeup on the `vecattr`/`expandos` facility forthcoming.)